### PR TITLE
Allow the sources of conformer generation failures to be retrieved

### PR DESCRIPTION
**Pull request by greglandrum**
The conformer generation code has a number of internal tests that it does to make sure the output conformers are reasonable. This PR makes it possible for client code to determine how often each of those tests fail.

This is mainly useful for debugging purposes, so it's not necessarily super well documented at the moment.

### CODE REVIEW

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp [hunk 1]
```diff
@@ -721,42 +756,91 @@ bool embedPoints(RDGeom::PointPtrVect *positions, detail::EmbedArgs eargs,
       std::cerr << "Initial embedding failed!, Iter: " << iter << std::endl;
```
Comment (sriniker) [inline Code/GraphMol/DistGeomHelpers/Embedder.cpp:756]: This is within #ifdef DEBUG_EMBEDDING, but it is still two "if (!gotCoords) {}" statements after each other (line 720 and 724). Can they maybe be combined?

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp [hunk 2]
```diff
@@ -721,42 +756,91 @@ bool embedPoints(RDGeom::PointPtrVect *positions, detail::EmbedArgs eargs,
...
+#endif
+            embedParams.failures[EmbedFailureCauses::ETK_MINIMIZATION]++;
+          }
+        }
       }
       // test if chirality is correct
       if (embedParams.enforceChirality && gotCoords &&
           (eargs.chiralCenters->size() > 0)) {
         gotCoords =
             EmbeddingOps::finalChiralChecks(positions, eargs, embedParams);
```
Comment (sriniker) [inline Code/GraphMol/DistGeomHelpers/Embedder.cpp:839]: Maybe add a comment that the tracking of the fails for this step is done within finalChiralChecks()

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp [hunk 3]
```diff
@@ -1091,7 +1175,14 @@ std::vector<std::vector<unsigned int>> getMolSelfMatches(
 }  // end of namespace detail
 
 void EmbedMultipleConfs(ROMol &mol, INT_VECT &res, unsigned int numConfs,
-                        const EmbedParameters &params) {
+                        EmbedParameters &params) {
+  if (params.trackFailures) {
+#ifdef RDK_BUILD_THREADSAFE_SSS
+    std::lock_guard<std::mutex> lock(GetFailMutex());
+#endif
+    params.failures.resize(10);
```
Comment (sriniker) [inline Code/GraphMol/DistGeomHelpers/Embedder.cpp:1183]: Is there a way to not hardcode the size of failures?

**Review by sriniker — COMMENTED** [review]
Looks good, just three comments/suggestions.

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/Embedder.cpp:756]: Even simpler: I just removed the debugging statement.

File: Code/GraphMol/DistGeomHelpers/Embedder.cpp (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/DistGeomHelpers/Embedder.cpp:1183]: We could either use a map or check the size of the failures vector each time it's used and resize if necessary.
Either seems kind of overkill.

To make it a bit more obvious that this exists (and what its value is), I'll make a constant `MAX_TRACKED_FAILURES`

Comment (sriniker) [reply Code/GraphMol/DistGeomHelpers/Embedder.cpp]: That makes sense

**Comment by greglandrum** [comment]
merging after @sriniker's review

**Pull request merged**